src/commands.rs

```diff
@@ -0,0 +1,104 @@
+
+use twitchchat::PrivmsgExt;
+
+// these are commands only the broadcaster can do
+mod broadcaster;
+
+// these are commands all users can do
+mod user;
+
+mod command;
+pub use command::Command;
+
+mod state;
+pub use state::State;
+
+fn split_user_cmd(data: &str) -> Option<(&str, Option<&str>)> {
+    let mut iter = data
+        .splitn(2, ' ')
+        .filter_map(|s| Some(s.trim()).filter(|c| !c.is_empty()));
+    Some((iter.next()?, iter.next()))
+}
+
+pub struct Commands {
+    state: State,
+}
+
+impl Commands {
+    /// Initialize the commands with the provided state
+    pub const fn new(state: State) -> Self {
+        Self { state }
+    }
+
+    /// Tries to dispatch this command
+    pub async fn dispatch(&mut self, command: &Command<'_>) -> Result<()> {
+        // broadcaster commands
+        if command.msg.is_broadcaster() {
+            self.elevated_dispatch(command).await?;
+        }
+        // the rest
+        self.normal_dispatch(command).await
+    }
+
+    // broadcaster commands
+    async fn elevated_dispatch(&mut self, command: &Command<'_>) -> Result<()> {
+        match (command.cmd, command.args) {
+            // '!quit'
+            ("quit", None) => return self.quit(command).await,
+            // '!unset cmd'
+            ("unset", Some(user_cmd)) => return self.unset(command, user_cmd).await,
+            // '!set cmd'
+            ("set", Some(user_cmd)) => {
+                // split 'user_cmd' into 'name' 'template here'?
+                let (user_cmd, template) = split_user_cmd(user_cmd).unwrap();
+                return self.set(command, user_cmd, template).await;
+            }
+            // '!give currency'
+            ("give", Some(user_cmd)) => {
+                // split 'user_cmd' into 'name' 'points'?
+                let (user_cmd, template) = split_user_cmd(user_cmd).unwrap();
+                return self.give(command, user_cmd, template).await;
+            }
+            // nothing else
+            _ => {}
+        }
+        Ok(())
+    }
+
+    // all user commands
+    async fn normal_dispatch(&mut self, command: &Command<'_>) -> Result<()> {
+        match (command.cmd, command.args) {
+            // '!commands' list
+            ("commands", None) => self.list_commands(command).await,
+            // <!user_command>
+            (user_cmd, None) => {
+                let (name, user_id) = (command.msg.name(), command.msg.user_id());
+                self.do_template(command, user_cmd, name, user_id).await
+            }
+            // invalid input
+            _ => Err(KappaError::BadInput(command.msg.data().to_string())),
+        }
+    }
+
+    // note: these are the same for now.
+    //
+    // helper which sends a response
+    fn send_response(&mut self, command: &Command<'_>, response: impl AsRef<str>) -> Result<()> {
+        // say just sends a message associated with the Privmsg
+        command.msg.say(&mut self.state.writer, response.as_ref())?;
```

Comment:
Why is `say()` on the PrivMsg and not off of something else? Is this a response to the sender of that command? I'd expect a reply to be `reply()`. `say()` makes sense if the response is going back to the channel, but it wasn't obvious to me at first glance.

Reply:
`say()` and `reply()` come from the `PrivmsgExt` trait. Its just a shorthand for doing 
```rust
// say
let output = commands::privmsg(msg.channel(), "foobar"); 
// reply -- this uses the new threaded system that is enabled in some chats. 
// if its not available then twitch will degrade to '@foo message' for you.
let output = commands::reply(msg.channel(), msg.id(), "foobar"); 
writer.encode(output);
```

I agree I need to document that trait. Currently its rather sparse. I need to highlight it in lib.rs.

The extension trait uses a sync writer under the hood so you don't need an async closure. It just encodes to a `Box<[u8]>` and sends it over a channel.